beets/metadata_plugins.py

```diff
@@ -0,0 +1,352 @@
+    def album_for_id(self, album_id: str) -> AlbumInfo | None:
+        """Return :py:class:`AlbumInfo` object or None if no matching release was
+        found."""
+        raise NotImplementedError
+
+    def tracks_for_ids(self, ids: Sequence[str]) -> Iterable[TrackInfo | None]:
+        """Batch lookup of track metadata for a list of track IDs.
+
+        Given a list of track identifiers, yields corresponding TrackInfo objects.
+        Missing tracks result in None values in the output iterator. Plugins may
+        implement this for optimized batched lookups instead of single calls to
+        track_for_id.
+        """
+
+        return iter(self.track_for_id(id) for id in ids)
+
+    @abc.abstractmethod
+    def track_for_id(self, track_id: str) -> TrackInfo | None:
+        """Return a :py:class:`AlbumInfo` object or None if no matching release was
+        found.
+        """
+        raise NotImplementedError
+
+    # ---------------------------------- search ---------------------------------- #
+
+    @abc.abstractmethod
+    def candidates(
+        self,
+        items: Sequence[Item],
+        artist: str,
+        album: str,
+        va_likely: bool,
+    ) -> Iterable[AlbumInfo]:
+        """Return :py:class:`AlbumInfo` candidates that match the given album.
+
+        Used in the autotag functionality to search for albums.
+
+        :param items: List of items in the album
+        :param artist: Album artist
+        :param album: Album name
+        :param va_likely: Whether the album is likely to be by various artists
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def item_candidates(
+        self, item: Item, artist: str, title: str
+    ) -> Iterable[TrackInfo]:
+        """Return :py:class:`TrackInfo` candidates that match the given track.
+
+        Used in the autotag functionality to search for tracks.
+
+        :param item: Track item
+        :param artist: Track artist
+        :param title: Track title
+        """
+        raise NotImplementedError
+
+    # --------------------------------- distances -------------------------------- #
+
+    def album_distance(
+        self,
+        items: Sequence[Item],
+        album_info: AlbumInfo,
+        mapping: dict[Item, TrackInfo],
+    ) -> Distance:
+        return _get_distance(
+            data_source=self.data_source, info=album_info, config=self.config
+        )
+
+    def track_distance(
+        self,
+        item: Item,
+        info: TrackInfo,
+    ) -> Distance:
+        return _get_distance(
+            data_source=self.data_source, info=info, config=self.config
+        )
+
+
+class IDResponse(TypedDict):
+    """Response from the API containing an ID."""
+
+    id: str
+
+
+class SearchFilter(TypedDict):
+    artist: NotRequired[str]
+    album: NotRequired[str]
+
+
+R = TypeVar("R", bound=IDResponse)
+
+
+class SearchApiMetadataSourcePlugin(
+    Generic[R], MetadataSourcePlugin, metaclass=abc.ABCMeta
+):
```

Comment:
Nice, I like the way you solved the issue of `_search_api` not being required for discogs and musicbrainz.